File: tests/scripts/requestsTests/putTests.cpp
#include "utils.cpp"

void	putTest(const std::string &url, const std::string &tabs)
{
	testServers("", url, tabs, "PUT");
}

void	putTests(void)
{
	printHeader("Allowed Case");
	putTest("/put/allowed/main.html", "\t\t\t\t\t");

	printHeader("Method Not Allowed Case");
	putTest("/put/notAllowed/main.html", "\t\t\t\t\t");

	printHeader("No Right on folder Case");
	putTest("/put/forbidden/main.html", "\t\t\t\t\t");

	printHeader("File Already existing With right Case");
	putTest("/put/mainRight.html", "\t\t\t\t\t");

	printHeader("File Already existing With no right Case");
	putTest("/put/mainNoRight.html", "\t\t\t\t\t");

	printHeader("Creating directory Case");
	putTest("/put/directory/", "\t\t\t\t\t");

	printHeader("Creating directory that already exists, with right and empty Case");
	putTest("/put/alreadyExistingDirEmpty/", "\t\t\t\t\t");

	printHeader("Creating directory that already exists, with right and not empty Case");
	putTest("/put/alreadyExistingDir/", "\t\t\t\t\t");

	printHeader("Creating directory that already exists, with no right and empty Case");
	putTest("/put/alreadyExistingDirEmptyNoRight/",  "\t\t\t\t\t");

	printHeader("Creating directory that already exists, with no right and not empty Case");
	putTest("/put/alreadyExistingDirNoRight/", "\t\t\t\t\t");
}

int	main(void)
{
	if (std::system("curl -I localhost:8181 > /dev/null 2>&1") != 0)
	{
		std::cout << "Nginx server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}
	if (std::system("curl -I localhost:8080 > /dev/null 2>&1") != 0)
	{
		std::cout << "Webserv server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}

	if (std::system("mkdir -p ../../../code/unitTest && cd ../../../code/unitTest && ../../tests/scripts/requestsTests/initPutTest.sh") != 0  // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/initPutTest.sh") != 0)  // For nginx
	{
		std::cout << "Error executing the put init scripts" << std::endl;
		return (EXIT_FAILURE);
	}

	putTests();


	if (std::system("cd ../../../code/unitTest && ../../tests/scripts/requestsTests/cleanPutTest.sh") != 0 // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/cleanPutTest.sh") != 0) // For nginx
	{
		std::cout << "Error executing the put cleaning scripts" << std::endl;
		return (EXIT_FAILURE);
	}
	return (EXIT_SUCCESS);
}

File: tests/scripts/requestsTests/putTests.mjs
import { compareGoodRequests } from "./compareRequests.mjs"
import { verifyServersAreRunning, exec, printHeader, generateString, COLOR_GREEN, COLOR_RED, COLOR_RESET } from "./utils.mjs"

let	failedTests = [];

async function	runGoodPutTest(target)
{
	const	result = await compareGoodRequests(target, "PUT", generateString(10, 100), {});
	if (result == true)
		console.log(COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		console.log(COLOR_RED + "[KO] " + COLOR_RESET);
		failedTests.push(target);
	}
	return (result);
}

async function runTests()
{
	printHeader("Allowed Case");
	await runGoodPutTest("/put/allowed/main.html");

	printHeader("Method Not Allowed Case");
	await runGoodPutTest("/put/notAllowed/main.html");

	printHeader("No Right on folder Case");
	await runGoodPutTest("/put/forbidden/main.html");

	printHeader("File Already existing With right Case");
	await runGoodPutTest("/put/mainRight.html");

	printHeader("File Already existing With no right Case");
	await runGoodPutTest("/put/mainNoRight.html");

	printHeader("Creating directory Case");
	await runGoodPutTest("/put/directory/");

	printHeader("Creating directory that already exists, with right and empty Case");
	await runGoodPutTest("/put/alreadyExistingDirEmpty/");

	printHeader("Creating directory that already exists, with right and not empty Case");
	await runGoodPutTest("/put/alreadyExistingDir/");

	printHeader("Creating directory that already exists, with no right and empty Case");
	await runGoodPutTest("/put/alreadyExistingDirEmptyNoRight/");

	printHeader("Creating directory that already exists, with no right and not empty Case");
	await runGoodPutTest("/put/alreadyExistingDirNoRight/");

	printHeader("Existing File");
	await runGoodPutTest("/put/allowed/main.html");

	printHeader("Folder Not Existing");
	await runGoodPutTest("/put/allowed/whereAmI/test.txt");


	exec("cd ../../webserv && ../scripts/requestsTests/cleanPutTest.sh"); // clean webserv
	exec("rm -rf ../../nginx/*"); // clean nginx

	exec("cd ../../webserv && ../scripts/requestsTests/initPutTest.sh"); // setup webserv
	exec("cp -pr ../../webserv/* ../../nginx/"); // setup nginx

	printHeader("Same But With Query String");
	await runGoodPutTest("/put/allowed/main.html?truc=var");
	await runGoodPutTest("/put/notAllowed/main.html?truc=var");
	await runGoodPutTest("/put/forbidden/main.html?truc=var");
	await runGoodPutTest("/put/mainRight.html?truc=var");
	await runGoodPutTest("/put/mainNoRight.html?truc=var");
	await runGoodPutTest("/put/directory/?truc=var");
	await runGoodPutTest("/put/alreadyExistingDirEmpty/?truc=var");
	await runGoodPutTest("/put/alreadyExistingDir/?truc=var");
	await runGoodPutTest("/put/alreadyExistingDirEmptyNoRight/?truc=var");
	await runGoodPutTest("/put/alreadyExistingDirNoRight/?truc=var");
	await runGoodPutTest("/put/allowed/main.html?truc=var");
	await runGoodPutTest("/put/allowed/whereAmI/test.txt?truc=var");

	if (failedTests.length == 0)
		printHeader("Everything Done : " + COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		printHeader("Everything Done : " + COLOR_RED + "[KO] " + COLOR_RESET);
		console.table(failedTests);
	}
}

async function	run()
{
	verifyServersAreRunning();

	exec("cd ../../webserv && ../scripts/requestsTests/initPutTest.sh"); // setup webserv
	exec("cp -pr ../../webserv/* ../../nginx/"); // setup nginx

	await runTests();

	exec("cd ../../webserv && ../scripts/requestsTests/cleanPutTest.sh"); // clean webserv
	exec("rm -rf ../../nginx/*"); // clean nginx
}

run();
